### PR TITLE
Various enhancements and improvements

### PR DESCRIPTION
**Pull request by jegnux**
Hello !

Here's a set of various enhancements and improvements I'd like to see in this library. 
I recognise that some changes are a little bit opinionated so feel free to discuss/reject some of them if needed.

Also, some changes are source breaking and I don't know what's your politic about it? Do you prefer we keep both API with a `deprecated` annotation? Or we can just keep the new version but we will eventually release this version under a new major version?

- Modernize API by using `by:` label instead of `of:` to match more closely with Swift stdlib
- Use `reduce(into:)` for better performances
Enhance `or` and `and` to immediatly return on the first element that doesn't match the expression requirement – it has the benefits to make it work with infinite sequence
- Add a specialized version of `distinct()` that works with `Hashable` elements in order to use a `Set` – it should make the `contains` lookup faster
- Enhance `filter` to immediatly return on the first element that doesn't match the expression requirement
- Pull NSPredicate init out of `filter` closure in order to init it only once
- Leverage Swift 4.0 new "subscript with default value" Dictionary API (SE-0165)
- Use stdlib `min(by:)` and `max(by:)` methods instead of a custom `reduce()`

### CODE REVIEW

**Comment by vincent-pradeilles** [comment]
Hello @jegnux 👋

Lots of great ideas here, thanks for implementing them 👍

There's one point that I'd like to discuss, it's the use of `by:` instead of `of:`:

When it's used in the context of a function like `func min(by:) -> Element`, I agree with its use, because we are retrieving an `Element` by the minimum of one of its properties.

But when it's a function like `func sum(by:) -> Double`, I find `of:` makes more sense, because we are directly returning the aggregate of the property, and not an `Element`.

What's your opinion on it?

**Comment by jegnux** [comment]
@vincent-pradeilles 
> But when it's a function like func sum(by:) -> Double, I find of: makes more sense, because we are directly returning the aggregate of the property, and not an Element.

Actually I agree, I might have **find/replace `(of:` by `(by:`** too fast 😅
Let me change this.

File: Sources/Operators/filterLike.swift [hunk 1]
```diff
@@ -10,6 +10,7 @@ import Foundation
 
 extension Sequence {
     public func filter(where attribute: KeyPath<Element, String>, like regex: String) -> [Element] {
-        return filter { return NSPredicate(format:"SELF MATCHES %@", regex).evaluate(with: $0[keyPath: attribute]) }
+        let predicate = NSPredicate(format:"SELF MATCHES %@", regex)
```
Comment (vincent-pradeilles) [inline Sources/Operators/filterLike.swift:13]: 👍

**Comment by vincent-pradeilles** [comment]
Merged and published on CocoaPods as version `1.4.0`!

Thanks for the contribution 👍

**Comment by jegnux** [comment]
@vincent-pradeilles awesome! Thanks

**Pull request merged**